src/App.jsx

```diff
@@ -1,35 +1,100 @@
-import { useState } from 'react'
-import reactLogo from './assets/react.svg'
-import viteLogo from '/vite.svg'
-import './App.css'
+import { useState, useEffect, useRef } from "react";
+import CharacterList from "./components/CharacterList";
+import Filter from "./components/Filter";
+import api from "./axiosInstance";
+import "./App.module.css";
 
 function App() {
-  const [count, setCount] = useState(0)
+  const [characters, setCharacters] = useState([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState(null);
+  const [currentPage, setCurrentPage] = useState(1);
+  const [hasMore, setHasMore] = useState(true);
+  const [filters, setFilters] = useState({
+    name: "",
+    status: "None",
+    gender: "None",
+  });
+
+  const observer = useRef(); 
+
+  const lastCharacterElementRef = useRef(null); 
+
+  const fetchCharacters = async (page = 1, newFilters = filters) => {
+    setIsLoading(true);
+    try {
+      const filterParams = {
+        page,
+        ...(newFilters.name && { name: newFilters.name }),
+        ...(newFilters.status !== "None" && { status: newFilters.status }),
+        ...(newFilters.gender !== "None" && { gender: newFilters.gender }),
+      };
+
+      const response = await api.get("/", { params: filterParams });
+
+      setCharacters((prevCharacters) => {
```

Comment:
lets store not only results but store all the data, received from API. And then you can get the value like "hasMore" and "characters" from that data